### PR TITLE
Extract Shared Assessment Layout from modeling assessment ui

### PR DESCRIPTION
**Pull request by jpbernius**
### Checklist
- [x] I tested the changes and all related features on the test server https://artemistest.ase.in.tum.de.
- [x] I documented my source code using the ~JavaDoc /~ JSDoc style.
- [x] ~I added integration test cases for the server (Spring) related to the features~
- [X] I added integration test cases for the client (Jest) related to the features
- [X] I added screenshots/screencast of my UI changes
- [x] ~I translated all the newly inserted strings~

### Motivation and Context
For updating the text assessment UI, we want to reuse existing UI from modeling assessment.

### Description
This PR extracts a majority of the modeling assessment layout into it's own component, so we can simply reuse it for text exercises. This *does not* modify the visual appearance!

### Steps for Testing

1. Participate with a modeling exercise (create before if necessary)
2. Assess your submission
3. Submit a complaint
4. Respond to the complaint
5. Find no differences to the same flow before this change.

### Screenshots
Screen Shot 2019-09-06 at 13 37 21

### CODE REVIEW

File: src/main/webapp/app/assessment-shared/assessment-header/assessment-header.component.ts [hunk 1]
```diff
@@ -1,26 +1,28 @@
...
 @Component({
     selector: 'jhi-assessment-header',
     templateUrl: './assessment-header.component.html',
     styleUrls: ['./assessment-header.component.scss'],
 })
-export class AssessmentHeaderComponent {
-    @Input() showBackButton: boolean;
-    @Output() goBack = new EventEmitter<void>();
+export class AssessmentHeaderComponent implements OnInit {
+    showBackButton: boolean;
```
Comment (thilo-behnke) [inline src/main/webapp/app/assessment-shared/assessment-header/assessment-header.component.ts]: At the moment the backButton is always shown. Is a scenario possible in which we don't want to show it?
This makes the component's inner workings less obvious as most parameters are passed through `@Input`s, while this one is passed through the URL.

This also implicity creates a bug, as the showBackButton prop is only taken from the URL on `ngOnInit`.

I would personally vote for removing it.

File: src/main/webapp/app/assessment-shared/assessment-header/assessment-header.component.ts [hunk 2]
```diff
@@ -33,4 +35,11 @@ export class AssessmentHeaderComponent {
     @Output() cancel = new EventEmitter<void>();
     @Output() resolveConflict = new EventEmitter<void>();
     @Output() nextSubmission = new EventEmitter<void>();
+
+    constructor(private accountService: AccountService, private route: ActivatedRoute) {}
+
+    ngOnInit() {
+        this.isAtLeastInstructor = this.accountService.hasAnyAuthorityDirect(['ROLE_ADMIN', 'ROLE_INSTRUCTOR']);
```
Comment (thilo-behnke) [inline src/main/webapp/app/assessment-shared/assessment-header/assessment-header.component.ts]: This does not make sure that the instructor is an instructor of this course.

Comment (thilo-behnke) [reply src/main/webapp/app/assessment-shared/assessment-header/assessment-header.component.ts]: Also it should probably be an input as the component will only be used in the `assessment-layout.component`. Then we could both remove the accountService and the route (see comment above), which would make the component dumb and easy to test.

File: src/main/webapp/app/assessment-shared/assessment-header/assessment-header.component.ts [hunk 3]
```diff
@@ -33,4 +35,11 @@ export class AssessmentHeaderComponent {
...
+        this.showBackButton = this.route.snapshot.queryParamMap.get('showBackButton') === 'true';
```
Comment (thilo-behnke) [inline src/main/webapp/app/assessment-shared/assessment-header/assessment-header.component.ts]: So to recap how this works:
- The router is used to access a component that has the `assessment-layout` embedded somewhere in it.
- Then the `assessment-header` component within the `assessment-layout` uses the url of the route to extract the showBackButton.

I think this is weird, because a client should only have to look at the top level component (here `assessment-layout`) to understand the configuration options. As mentioned above it should be and input or avoided as it seems to be true for all clients atm.

File: src/test/javascript/spec/helpers/mutable.ts [hunk 4]
```diff
@@ -0,0 +1,3 @@
+export type Mutable<T> = {
```
Comment (thilo-behnke) [inline src/test/javascript/spec/helpers/mutable.ts]: I struggled to understand this :sweat_smile: could you maybe provide some docs to what this type does?
Doesn't `readonly` indicate that it is immutable?

Comment (jpbernius) [reply src/test/javascript/spec/helpers/mutable.ts]: Yes, this removes the readonly, therefore the **`-`**`readonly`.
I will add a documentation for this.

File: src/main/webapp/app/assessment-shared/assessment-header/assessment-header.component.ts (hunk 1 above)
Comment (jpbernius) [reply src/main/webapp/app/assessment-shared/assessment-header/assessment-header.component.ts]: We do not show the back button if you open an Assessment from the "Assessment Dashboard"

<details><summary>Screenshot</summary>
Screen Shot 2019-09-10 at 10 42 54
</details>

Comment (jpbernius) [reply src/main/webapp/app/assessment-shared/assessment-header/assessment-header.component.ts]: Changed. ✅

File: src/main/webapp/app/assessment-shared/assessment-header/assessment-header.component.ts (hunk 3 above)
Comment (jpbernius) [reply src/main/webapp/app/assessment-shared/assessment-header/assessment-header.component.ts]: Refactored this.

File: src/main/webapp/app/assessment-shared/assessment-header/assessment-header.component.ts (hunk 2 above)
Comment (jpbernius) [reply src/main/webapp/app/assessment-shared/assessment-header/assessment-header.component.ts]: Done.

File: src/main/webapp/app/assessment-shared/assessment-layout/assessment-layout.component.html [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+        [isAssessor]="isAssessor"
+        [isAtLeastInstructor]="isAtLeastInstructor"
+        [canOverride]="canOverride"
+        [assessmentsAreValid]="assessmentsAreValid"
+        [result]="result"
+        (save)="save.emit()"
+        (submit)="submit.emit()"
+        (cancel)="cancel.emit()"
+        (nextSubmission)="nextSubmission.emit()"
+        [hasConflict]="conflicts != null"
```
Comment (maxr96) [inline src/main/webapp/app/assessment-shared/assessment-layout/assessment-layout.component.html]: Please try to avoid `!=` and use `!==` to avoid casting and strictly check equality

File: src/main/webapp/app/modeling-assessment-editor/modeling-assessment-editor.component.ts [hunk 6]
```diff
@@ -84,22 +81,20 @@ export class ModelingAssessmentEditorComponent implements OnInit, OnDestroy {
         });
         this.isAtLeastInstructor = this.accountService.hasAnyAuthorityDirect(['ROLE_ADMIN', 'ROLE_INSTRUCTOR']);
 
-        this.route.params.subscribe(params => {
-            const submissionId: String = params['submissionId'];
-            const exerciseId = Number(params['exerciseId']);
+        this.route.paramMap.subscribe(params => {
+            const submissionId: String | null = params.get('submissionId');
```
Comment (maxr96) [inline src/main/webapp/app/modeling-assessment-editor/modeling-assessment-editor.component.ts]: In this case if `submissionId` is null we will call `loadSubmission`, which doesn't expect null. Is this intended?

File: src/main/webapp/app/assessment-shared/assessment-layout/assessment-layout.component.ts [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+import { ComplaintResponse } from 'app/entities/complaint-response';
+
+/**
+ * The <jhi-assessment-layout> component provides the basic layout for an assessment page.
+ * It shows the header, alerts for complaints on top and the complaint form at the bottom of the page.
+ * The actual assessment needs to be inserted using content projection.
+ * Components using this component need to provide Inputs and handle Outputs. This component does not perform assessment logic.
+ */
+@Component({
+    selector: 'jhi-assessment-layout',
```
Comment (maxr96) [inline src/main/webapp/app/assessment-shared/assessment-layout/assessment-layout.component.ts]: This is totally unrelated to this PR. But I am curious why we actually name our components with `jhi-..`. Why not use `artemis-..` to not conflict with `jhi` components like `jhi-alert..`? Maybe we can start naming our newly added generic components this way?

**Comment by SiggZ** [comment]
> * This is also not related to this specific PR, but I am able to submit score, that is more than maximal and this value is saved in Modeling Assessment Editor (even if I exit and open it again). On the tutor exercise dashboard it is saved properly. But maybe we should warn tutor that he exceeded the maximum amount of points.

@maxr96: this is the intended behavior. Tutors are allowed to submit scores higher than the max score, for the student the displayed score is limited by the max score though. A warning about exceeding the max score might be a good idea, but it is out of scope of this PR.

File: src/main/webapp/app/assessment-shared/assessment-layout/assessment-layout.component.ts (hunk 7 above)
Comment (thilo-behnke) [reply src/main/webapp/app/assessment-shared/assessment-layout/assessment-layout.component.ts]: I agree, we have defined this in our angular.json file:
https://github.com/ls1intum/Artemis/blob/develop/angular.json#L57

This was probably just an oversight. It would be nice if we could define this **per feature module** (e.g. participation- or code-editor-). https://angular.io/guide/styleguide#style-02-07
It would be nice if we could move this discussion to a Github issue.

**Comment by jpbernius** [comment]
> Since you already try to generalize the whole layout, it would be awesome if we have the same header as for text exercise. This is what I planned to do anyway.
> -- @maxr96 

Currently, we are planning to move away from the large header in the text exercises a well, as this is mostly wasting space.

> This issue is probably not really related to this PR, but when I try to add feedback for class the dialog is a bit cut off. This works properly for Chrome, but not Firefox (both tested on Linux)
> -- @maxr96 

If this is still the case, we should raise a separate issue for this.

> Note, that this PR will conflict with #794 where I change the query parameter showBackButton to hideBackButton.
> -- @SiggZ 

I updated this PR with your changes.

File: src/main/webapp/app/modeling-assessment-editor/modeling-assessment-editor.component.ts (hunk 6 above)
Comment (jpbernius) [reply src/main/webapp/app/modeling-assessment-editor/modeling-assessment-editor.component.ts]: This is the same behavior we had before.
The null case is "handled" by the cast in Line 90:
```ts
this.loadSubmission(Number(submissionId));
```
Which converts `null` to `0`.

File: src/main/webapp/app/assessment-shared/assessment-layout/assessment-layout.component.ts (hunk 7 above)
Comment (jpbernius) [reply src/main/webapp/app/assessment-shared/assessment-layout/assessment-layout.component.ts]: I crated #843 for this discussion.
